### PR TITLE
POST API

### LINKED ISSUE #442: Create POST APIs
See https://github.com/fossasia/open-event-orga-server/issues/336 for the schema

Currently these POST methods will be publicy accessible (login needed). Only after #429 is done (which in on hold currently), the access control can be added to it.

### PR DESCRIPTION
**Pull request by aviaryan**
Fixes #442 and #462
Includes PR #456 

Currently no ACL has been implemented in this patch although login is required to be able to update data on server. Using Python requests - 
```python
from requests import post 
url = 'http://localhost:5000/api/v2/events'
print post(url + '/1/formats', json={'label_en': 'eng', 'name': 'eng'}, auth=('username', 'password')).json()
```
In swagger UI, you need to be signed in to open event before you can make POST requests. :sunglasses: 

#### A question

@juslee @leto @shivamMg 

For POST request to create a new session, I am representing related models only by their ids. like `level_id` instead of full level object (below)
```json
"level": {
    "id": 1,
    "label_en": "string"
  },
```

The current schema to send a POST req to create session is as follows. As you can see, I am using `level_id`, `track_id` and so..
```json
{
  "abstract": "string",
  "description": "string",
  "end_time": "2016-05-29T12:44:03.743Z",
  "language_id": 0,
  "level_id": 0,
  "microlocation_id": 0,
  "speakers": [
    {
       "id": 0,
       "name": "string"
    }
  ],
  "start_time": "2016-05-29T12:44:03.743Z",
  "subtitle": "string",
  "title": "string",
  "track_id": 0
}
```

Is using IDs better since when creating a session, if you want to attach a level, you need to have the id of level before hand ?
Should I change `speakers` to `speaker_ids`, a list of speaker id's ?

**UPDATE** - I have changed `speakers` to `speaker_ids` now.

### CODE REVIEW

File: open_event/api/speakers.py [hunk 1]
```diff
@@ -34,6 +32,17 @@
     'results': fields.List(fields.Nested(SPEAKER))
 })
 
+SPEAKER_POST = api.clone('SpeakerPost', SPEAKER)
+del SPEAKER_POST['id']
+del SPEAKER_POST['sessions']  # don't allow adding sessions
+
+
+# Create DAO
+class SpeakerDAO(ServiceDAO):
```
Comment (juslee) [inline open_event/api/speakers.py]: Codacy Issue found: [Too few public methods (0/2)](https://www.codacy.com/app/triplez/open-event-orga-server/file/2651459026/issues/source?bid=3233864&fileBranchId=3324785#l41)

File: open_event/api/helpers.py [hunk 2]
```diff
@@ -121,3 +127,59 @@ def get_paginated_list(klass, url, args={}, **kwargs):
...
+    and save it to database
+    """
+    data['event_id'] = event_id
+    new_model = model(**data)
+    save_to_db(new_model, "Model %s saved" % model.__name__)
+    update_version(event_id, False, "session_ver")
+    return new_model
+
+
+def requires_auth(f):
```
Comment (juslee) [inline open_event/api/helpers.py]: Codacy Issue found: [Invalid argument name "f"](https://www.codacy.com/app/triplez/open-event-orga-server/file/2651459021/issues/source?bid=3233864&fileBranchId=3324785#l154)

File: open_event/api/microlocations.py [hunk 3]
```diff
@@ -21,6 +19,16 @@
     'results': fields.List(fields.Nested(MICROLOCATION))
 })
 
+MICROLOCATION_POST = api.clone('MicrolocationPost', MICROLOCATION)
+del MICROLOCATION_POST['id']
+
+
+# Create DAO
+class MicrolocationDAO(ServiceDAO):
```
Comment (juslee) [inline open_event/api/microlocations.py]: Codacy Issue found: [Too few public methods (0/2)](https://www.codacy.com/app/triplez/open-event-orga-server/file/2651459024/issues/source?bid=3233864&fileBranchId=3324785#l27)

File: open_event/api/sponsors.py [hunk 4]
```diff
@@ -19,6 +17,16 @@
     'results': fields.List(fields.Nested(SPONSOR))
 })
 
+SPONSOR_POST = api.clone('SponsorPost', SPONSOR)
+del SPONSOR_POST['id']
+
+
+# Create DAO
+class SponsorDAO(ServiceDAO):
```
Comment (juslee) [inline open_event/api/sponsors.py]: Codacy Issue found: [Too few public methods (0/2)](https://www.codacy.com/app/triplez/open-event-orga-server/file/2651459027/issues/source?bid=3233864&fileBranchId=3324785#l25)

File: open_event/api/formats.py [hunk 5]
```diff
@@ -19,6 +17,16 @@
     'results': fields.List(fields.Nested(FORMAT))
 })
 
+FORMAT_POST = api.clone('FormatPost', FORMAT)
+del FORMAT_POST['id']
+
+
+# Create DAO
+class FormatDAO(ServiceDAO):
```
Comment (juslee) [inline open_event/api/formats.py]: Codacy Issue found: [Too few public methods (0/2)](https://www.codacy.com/app/triplez/open-event-orga-server/file/2651459020/issues/source?bid=3233864&fileBranchId=3324785#l25)

File: open_event/api/tracks.py [hunk 6]
```diff
@@ -25,6 +23,17 @@
     'results': fields.List(fields.Nested(TRACK))
 })
 
+TRACK_POST = api.clone('TrackPost', TRACK)
+del TRACK_POST['id']
+del TRACK_POST['sessions']
+
+
+# Create DAO
+class TrackDAO(ServiceDAO):
```
Comment (juslee) [inline open_event/api/tracks.py]: Codacy Issue found: [Too few public methods (0/2)](https://www.codacy.com/app/triplez/open-event-orga-server/file/2651459028/issues/source?bid=3233864&fileBranchId=3324785#l32)

File: open_event/api/levels.py [hunk 7]
```diff
@@ -18,6 +16,16 @@
     'results': fields.List(fields.Nested(LEVEL))
 })
 
+LEVEL_POST = api.clone('LevelPost', LEVEL)
+del LEVEL_POST['id']
+
+
+# Create DAO
+class LevelDAO(ServiceDAO):
```
Comment (juslee) [inline open_event/api/levels.py]: Codacy Issue found: [Too few public methods (0/2)](https://www.codacy.com/app/triplez/open-event-orga-server/file/2651459023/issues/source?bid=3233864&fileBranchId=3324785#l24)

File: open_event/api/languages.py [hunk 8]
```diff
@@ -19,6 +17,16 @@
     'results': fields.List(fields.Nested(LANGUAGE))
 })
 
+LANGUAGE_POST = api.clone('LanguagePost', LANGUAGE)
+del LANGUAGE_POST['id']
+
+
+# Create DAO
+class LanguageDAO(ServiceDAO):
```
Comment (juslee) [inline open_event/api/languages.py]: Codacy Issue found: [Too few public methods (0/2)](https://www.codacy.com/app/triplez/open-event-orga-server/file/2651459022/issues/source?bid=3233864&fileBranchId=3324785#l25)

File: open_event/api/sessions.py [hunk 9]
```diff
@@ -53,6 +58,42 @@
...
+})
+del SESSION_POST['id']
+del SESSION_POST['track']
+del SESSION_POST['level']
+del SESSION_POST['language']
+del SESSION_POST['microlocation']
+
+
+# Create DAO
+class SessionDAO(ServiceDAO):
```
Comment (juslee) [inline open_event/api/sessions.py]: Codacy Issue found: [Too few public methods (1/2)](https://www.codacy.com/app/triplez/open-event-orga-server/file/2651459025/issues/source?bid=3233864&fileBranchId=3324785#l75)

File: open_event/api/languages.py (hunk 8 above)
Comment (juslee) [inline open_event/api/languages.py]: Codacy Issue found: [Too few public methods (0/2)](https://www.codacy.com/app/triplez/open-event-orga-server/file/2651468186/issues/source?bid=3233864&fileBranchId=3324785#l25)

File: open_event/api/speakers.py (hunk 1 above)
Comment (juslee) [inline open_event/api/speakers.py]: Codacy Issue found: [Too few public methods (0/2)](https://www.codacy.com/app/triplez/open-event-orga-server/file/2651468190/issues/source?bid=3233864&fileBranchId=3324785#l41)

File: open_event/api/helpers.py (hunk 2 above)
Comment (juslee) [inline open_event/api/helpers.py]: Codacy Issue found: [Invalid argument name "f"](https://www.codacy.com/app/triplez/open-event-orga-server/file/2651468185/issues/source?bid=3233864&fileBranchId=3324785#l154)

File: open_event/api/sessions.py [hunk 10]
```diff
@@ -46,13 +57,54 @@
...
+del SESSION_POST['id']
+del SESSION_POST['track']
+del SESSION_POST['level']
+del SESSION_POST['language']
+del SESSION_POST['microlocation']
+del SESSION_POST['format']
+
+
+# Create DAO
+class SessionDAO(ServiceDAO):
```
Comment (juslee) [inline open_event/api/sessions.py]: Codacy Issue found: [Too few public methods (1/2)](https://www.codacy.com/app/triplez/open-event-orga-server/file/2651468189/issues/source?bid=3233864&fileBranchId=3324785#l84)

File: open_event/api/microlocations.py (hunk 3 above)
Comment (juslee) [inline open_event/api/microlocations.py]: Codacy Issue found: [Too few public methods (0/2)](https://www.codacy.com/app/triplez/open-event-orga-server/file/2651468188/issues/source?bid=3233864&fileBranchId=3324785#l27)

File: open_event/api/tracks.py (hunk 6 above)
Comment (juslee) [inline open_event/api/tracks.py]: Codacy Issue found: [Too few public methods (0/2)](https://www.codacy.com/app/triplez/open-event-orga-server/file/2651468192/issues/source?bid=3233864&fileBranchId=3324785#l32)

File: open_event/api/levels.py (hunk 7 above)
Comment (juslee) [inline open_event/api/levels.py]: Codacy Issue found: [Too few public methods (0/2)](https://www.codacy.com/app/triplez/open-event-orga-server/file/2651468187/issues/source?bid=3233864&fileBranchId=3324785#l24)

File: open_event/api/sponsors.py (hunk 4 above)
Comment (juslee) [inline open_event/api/sponsors.py]: Codacy Issue found: [Too few public methods (0/2)](https://www.codacy.com/app/triplez/open-event-orga-server/file/2651468191/issues/source?bid=3233864&fileBranchId=3324785#l25)

File: open_event/api/formats.py (hunk 5 above)
Comment (juslee) [inline open_event/api/formats.py]: Codacy Issue found: [Too few public methods (0/2)](https://www.codacy.com/app/triplez/open-event-orga-server/file/2651468184/issues/source?bid=3233864&fileBranchId=3324785#l25)

**Comment by juslee** [comment]
@aviaryan did you write your tests?

**Comment by aviaryan** [comment]
@juslee will write tests and update the PR

File: open_event/api/microlocations.py (hunk 3 above)
Comment (juslee) [inline open_event/api/microlocations.py]: Codacy Issue found: [Too few public methods (0/2)](https://www.codacy.com/app/triplez/open-event-orga-server/file/2652871195/issues/source?bid=3233864&fileBranchId=3324785#l27)

File: open_event/api/levels.py (hunk 7 above)
Comment (juslee) [inline open_event/api/levels.py]: Codacy Issue found: [Too few public methods (0/2)](https://www.codacy.com/app/triplez/open-event-orga-server/file/2652871194/issues/source?bid=3233864&fileBranchId=3324785#l24)

File: open_event/api/tracks.py (hunk 6 above)
Comment (juslee) [inline open_event/api/tracks.py]: Codacy Issue found: [Too few public methods (0/2)](https://www.codacy.com/app/triplez/open-event-orga-server/file/2652871199/issues/source?bid=3233864&fileBranchId=3324785#l32)

File: open_event/api/formats.py (hunk 5 above)
Comment (juslee) [inline open_event/api/formats.py]: Codacy Issue found: [Too few public methods (0/2)](https://www.codacy.com/app/triplez/open-event-orga-server/file/2652871191/issues/source?bid=3233864&fileBranchId=3324785#l25)

File: open_event/api/sponsors.py (hunk 4 above)
Comment (juslee) [inline open_event/api/sponsors.py]: Codacy Issue found: [Too few public methods (0/2)](https://www.codacy.com/app/triplez/open-event-orga-server/file/2652871198/issues/source?bid=3233864&fileBranchId=3324785#l25)

File: open_event/api/helpers.py (hunk 2 above)
Comment (juslee) [inline open_event/api/helpers.py]: Codacy Issue found: [Invalid argument name "f"](https://www.codacy.com/app/triplez/open-event-orga-server/file/2652871192/issues/source?bid=3233864&fileBranchId=3324785#l154)

File: open_event/api/speakers.py (hunk 1 above)
Comment (juslee) [inline open_event/api/speakers.py]: Codacy Issue found: [Too few public methods (0/2)](https://www.codacy.com/app/triplez/open-event-orga-server/file/2652871197/issues/source?bid=3233864&fileBranchId=3324785#l41)

File: open_event/api/languages.py (hunk 8 above)
Comment (juslee) [inline open_event/api/languages.py]: Codacy Issue found: [Too few public methods (0/2)](https://www.codacy.com/app/triplez/open-event-orga-server/file/2652871193/issues/source?bid=3233864&fileBranchId=3324785#l25)

File: open_event/api/sessions.py [hunk 11]
```diff
@@ -46,13 +57,56 @@
...
+del SESSION_POST['track']
+del SESSION_POST['speakers']
+del SESSION_POST['level']
+del SESSION_POST['language']
+del SESSION_POST['microlocation']
+del SESSION_POST['format']
+
+
+# Create DAO
+class SessionDAO(ServiceDAO):
```
Comment (juslee) [inline open_event/api/sessions.py]: Codacy Issue found: [Too few public methods (1/2)](https://www.codacy.com/app/triplez/open-event-orga-server/file/2652871196/issues/source?bid=3233864&fileBranchId=3324785#l86)

File: open_event/api/languages.py (hunk 8 above)
Comment (juslee) [inline open_event/api/languages.py]: Codacy Issue found: [Too few public methods (0/2)](https://www.codacy.com/app/triplez/open-event-orga-server/file/2653967394/issues/source?bid=3233864&fileBranchId=3324785#l25)

File: open_event/api/speakers.py (hunk 1 above)
Comment (juslee) [inline open_event/api/speakers.py]: Codacy Issue found: [Too few public methods (0/2)](https://www.codacy.com/app/triplez/open-event-orga-server/file/2653967398/issues/source?bid=3233864&fileBranchId=3324785#l41)

File: open_event/api/helpers.py (hunk 2 above)
Comment (juslee) [inline open_event/api/helpers.py]: Codacy Issue found: [Invalid argument name "f"](https://www.codacy.com/app/triplez/open-event-orga-server/file/2653967393/issues/source?bid=3233864&fileBranchId=3324785#l154)

File: open_event/api/sessions.py (hunk 11 above)
Comment (juslee) [inline open_event/api/sessions.py]: Codacy Issue found: [Too few public methods (1/2)](https://www.codacy.com/app/triplez/open-event-orga-server/file/2653967397/issues/source?bid=3233864&fileBranchId=3324785#l86)

File: open_event/api/formats.py (hunk 5 above)
Comment (juslee) [inline open_event/api/formats.py]: Codacy Issue found: [Too few public methods (0/2)](https://www.codacy.com/app/triplez/open-event-orga-server/file/2653967392/issues/source?bid=3233864&fileBranchId=3324785#l25)

File: open_event/api/tracks.py (hunk 6 above)
Comment (juslee) [inline open_event/api/tracks.py]: Codacy Issue found: [Too few public methods (0/2)](https://www.codacy.com/app/triplez/open-event-orga-server/file/2653967400/issues/source?bid=3233864&fileBranchId=3324785#l32)

File: tests/api/test_post_api.py [hunk 12]
```diff
@@ -0,0 +1,81 @@
...
+        self.assertEqual(401, response.status_code, msg=response.data)
+        # login and send the request again
+        self.login_user()
+        response = self.app.post(
+            path,
+            data=json.dumps(data),
+            headers={'content-type': 'application/json'}
+        )
+        self.assertEqual(200, response.status_code, msg=response.data)
+        self.assertIn('Test' + str(name).title(), msg=response.data)
```
Comment (juslee) [inline tests/api/test_post_api.py]: Codacy Issue found: [No value for argument 'container' in method call](https://www.codacy.com/app/triplez/open-event-orga-server/file/2653967465/issues/source?bid=3233864&fileBranchId=3324785#l51)

File: open_event/api/microlocations.py (hunk 3 above)
Comment (juslee) [inline open_event/api/microlocations.py]: Codacy Issue found: [Too few public methods (0/2)](https://www.codacy.com/app/triplez/open-event-orga-server/file/2653967396/issues/source?bid=3233864&fileBranchId=3324785#l27)

File: tests/api/test_post_api.py [hunk 13]
```diff
@@ -0,0 +1,81 @@
+import unittest
+import json
+
+from tests.setup_database import Setup
+from tests.utils import OpenEventTestCase
+from tests.auth_helper import login, register
+from tests.api.utils import create_event, get_path
+from tests.api.utils_post_data import *
```
Comment (juslee) [inline tests/api/test_post_api.py]: Codacy Issue found: [Wildcard import tests.api.utils_post_data](https://www.codacy.com/app/triplez/open-event-orga-server/file/2653967465/issues/source?bid=3233864&fileBranchId=3324785#l8)

File: open_event/api/sponsors.py (hunk 4 above)
Comment (juslee) [inline open_event/api/sponsors.py]: Codacy Issue found: [Too few public methods (0/2)](https://www.codacy.com/app/triplez/open-event-orga-server/file/2653967399/issues/source?bid=3233864&fileBranchId=3324785#l25)

File: tests/api/test_post_api.py (hunk 13 above)
Comment (juslee) [inline tests/api/test_post_api.py]: Codacy Issue found: [Wildcard import tests.api.utils_post_data](https://www.codacy.com/app/triplez/open-event-orga-server/file/2654023351/issues/source?bid=3233864&fileBranchId=3324785#l8)

File: open_event/api/sponsors.py (hunk 4 above)
Comment (juslee) [inline open_event/api/sponsors.py]: Codacy Issue found: [Too few public methods (0/2)](https://www.codacy.com/app/triplez/open-event-orga-server/file/2654023285/issues/source?bid=3233864&fileBranchId=3324785#l25)

File: open_event/api/helpers.py (hunk 2 above)
Comment (juslee) [inline open_event/api/helpers.py]: Codacy Issue found: [Invalid argument name "f"](https://www.codacy.com/app/triplez/open-event-orga-server/file/2654023279/issues/source?bid=3233864&fileBranchId=3324785#l154)

File: open_event/api/levels.py (hunk 7 above)
Comment (juslee) [inline open_event/api/levels.py]: Codacy Issue found: [Too few public methods (0/2)](https://www.codacy.com/app/triplez/open-event-orga-server/file/2654023281/issues/source?bid=3233864&fileBranchId=3324785#l24)

File: open_event/api/speakers.py (hunk 1 above)
Comment (juslee) [inline open_event/api/speakers.py]: Codacy Issue found: [Too few public methods (0/2)](https://www.codacy.com/app/triplez/open-event-orga-server/file/2654023284/issues/source?bid=3233864&fileBranchId=3324785#l41)

File: open_event/api/languages.py (hunk 8 above)
Comment (juslee) [inline open_event/api/languages.py]: Codacy Issue found: [Too few public methods (0/2)](https://www.codacy.com/app/triplez/open-event-orga-server/file/2654023280/issues/source?bid=3233864&fileBranchId=3324785#l25)

File: open_event/api/sessions.py (hunk 11 above)
Comment (juslee) [inline open_event/api/sessions.py]: Codacy Issue found: [Too few public methods (1/2)](https://www.codacy.com/app/triplez/open-event-orga-server/file/2654023283/issues/source?bid=3233864&fileBranchId=3324785#l86)

File: open_event/api/formats.py (hunk 5 above)
Comment (juslee) [inline open_event/api/formats.py]: Codacy Issue found: [Too few public methods (0/2)](https://www.codacy.com/app/triplez/open-event-orga-server/file/2654023278/issues/source?bid=3233864&fileBranchId=3324785#l25)

File: open_event/api/tracks.py (hunk 6 above)
Comment (juslee) [inline open_event/api/tracks.py]: Codacy Issue found: [Too few public methods (0/2)](https://www.codacy.com/app/triplez/open-event-orga-server/file/2654023286/issues/source?bid=3233864&fileBranchId=3324785#l32)

File: open_event/api/microlocations.py (hunk 3 above)
Comment (juslee) [inline open_event/api/microlocations.py]: Codacy Issue found: [Too few public methods (0/2)](https://www.codacy.com/app/triplez/open-event-orga-server/file/2654023282/issues/source?bid=3233864&fileBranchId=3324785#l27)

**Comment by aviaryan** [comment]
@juslee @shivamMg I have written the tests. Please review and merge.

File: open_event/api/speakers.py (hunk 1 above)
Comment (juslee) [inline open_event/api/speakers.py]: Codacy Issue found: [Too few public methods (0/2)](https://www.codacy.com/app/triplez/open-event-orga-server/file/2654653406/issues/source?bid=3233864&fileBranchId=3324785#l41)

File: tests/api/test_post_api.py [hunk 14]
```diff
@@ -0,0 +1,88 @@
...
```
Comment (juslee) [inline tests/api/test_post_api.py]: Codacy Issue found: [Wildcard import tests.api.utils_post_data](https://www.codacy.com/app/triplez/open-event-orga-server/file/2654653473/issues/source?bid=3233864&fileBranchId=3324785#l8)

File: open_event/api/languages.py (hunk 8 above)
Comment (juslee) [inline open_event/api/languages.py]: Codacy Issue found: [Too few public methods (0/2)](https://www.codacy.com/app/triplez/open-event-orga-server/file/2654653402/issues/source?bid=3233864&fileBranchId=3324785#l25)

File: open_event/api/helpers.py (hunk 2 above)
Comment (juslee) [inline open_event/api/helpers.py]: Codacy Issue found: [Invalid argument name "f"](https://www.codacy.com/app/triplez/open-event-orga-server/file/2654653401/issues/source?bid=3233864&fileBranchId=3324785#l154)

File: open_event/api/sessions.py (hunk 11 above)
Comment (juslee) [inline open_event/api/sessions.py]: Codacy Issue found: [Too few public methods (1/2)](https://www.codacy.com/app/triplez/open-event-orga-server/file/2654653405/issues/source?bid=3233864&fileBranchId=3324785#l86)

File: open_event/api/tracks.py (hunk 6 above)
Comment (juslee) [inline open_event/api/tracks.py]: Codacy Issue found: [Too few public methods (0/2)](https://www.codacy.com/app/triplez/open-event-orga-server/file/2654653408/issues/source?bid=3233864&fileBranchId=3324785#l32)

File: open_event/api/microlocations.py (hunk 3 above)
Comment (juslee) [inline open_event/api/microlocations.py]: Codacy Issue found: [Too few public methods (0/2)](https://www.codacy.com/app/triplez/open-event-orga-server/file/2654653404/issues/source?bid=3233864&fileBranchId=3324785#l27)

File: open_event/api/events.py [hunk 15]
```diff
@@ -45,6 +51,27 @@ def get(self):
...
         return get_object_list(EventModel)
 
+    @requires_auth
+    @api.doc('create_event')
+    @api.marshal_with(EVENT)
+    @api.expect(EVENT_POST, validate=True)
+    def post(self):
+        """Create an event"""
+        new_event = EventModel(**self.api.payload)
+        a = EventsUsers()
```
Comment (juslee) [inline open_event/api/events.py]: Codacy Issue found: [Invalid variable name "a"](https://www.codacy.com/app/triplez/open-event-orga-server/file/2654653399/issues/source?bid=3233864&fileBranchId=3324785#l61)

File: open_event/api/formats.py (hunk 5 above)
Comment (juslee) [inline open_event/api/formats.py]: Codacy Issue found: [Too few public methods (0/2)](https://www.codacy.com/app/triplez/open-event-orga-server/file/2654653400/issues/source?bid=3233864&fileBranchId=3324785#l25)

File: open_event/api/sponsors.py (hunk 4 above)
Comment (juslee) [inline open_event/api/sponsors.py]: Codacy Issue found: [Too few public methods (0/2)](https://www.codacy.com/app/triplez/open-event-orga-server/file/2654653407/issues/source?bid=3233864&fileBranchId=3324785#l25)

**Comment by aviaryan** [comment]
@juslee This is ready for merging

**Pull request merged**